iree/compiler/Dialect/HAL/Conversion/FlowToHAL/ConvertFlowToHAL.cpp

```diff
@@ -71,7 +73,14 @@ class ConvertFlowToHALPass
     : public PassWrapper<ConvertFlowToHALPass, OperationPass<ModuleOp>> {
  public:
   void getDependentDialects(DialectRegistry &registry) const override {
-    registry.insert<IREE::HAL::HALDialect>();
+    registry.insert<
+        IREE::HAL::HALDialect,
+        // The context is not yet available here and runOnOperation is too late
+        // to load the dialects. Therefore we have to register all dialects that
+        // the conversion patterns might create.
+        // TODO(#2958): Register dialects dynamically based on the
+        // HALConversionDialectInterface that are available.
+        IREE::Strings::StringsDialect, IREE::TensorList::TensorListDialect>();
```

Comment:
This is unfortunate - why is it needed? The interface is used to find/work with dialects that are already registered by someone else - if they aren't registered, they aren't needed. For example, if you are iree-opt and don't have the tensorlist dialect pulled in then you won't be able to work with tensorlist - that's expected - but if you are iree-tf-opt and do have it then this pass should find it.

Reply:
Even if these dialects are registered, we have to tell the pass manager to load them before running the pass pipeline. It uses the declared dependent dialects of the pass to do this (https://github.com/llvm/llvm-project/blob/master/mlir/lib/Pass/Pass.cpp#L749). This pass creates entities from these dialects if it sees the corresponding tf_strings and tf_tensorlist dialects. Per the TODO here, I don't see another way to do this without some changes to MLIR.

Reply:
![image](https://user-images.githubusercontent.com/75337/91881473-6380dc80-ec36-11ea-96f0-fcb140c1cf68.png)

This absolutely needs to be fixed in MLIR - is there an upstream bug for it? This is a full inversion of dependencies and like this there is no way to add an out-of-tree custom dialect that lowers into the HAL. That's really bad. If the next thing done is fixing this I'm ok with it, but it's a really bad change and it's not good that this was forced by upstream without a workaround.

Reply:
@joker-eph could we have a bug to track providing some way of allowing dynamic dialect dependencies for a pass?

Reply:
That sounds to me right now like a problem to solve within IREE's infrastructure more than with MLIR? It isn't clear to me what kind of facility are missing from upstream right now.

Reply:
The issue is that we did solve it in IREE's infrastructure but these latest changes broke that solution and (seemingly) prevent it from being possible to solve at all. So changes are required upstream to enable this scenario (which in a multi-level IR system are pretty core to the design).

Reply:
> (seemingly) prevent it from being possible to solve at all. 

I don't quite understand what is complex in solving this right now? I need to get more educated on the problem maybe, is there a description of the requirements for this pass?

Reply:
Right now with the global registry, we have a way for a out-of-tree user to say "here's how to lower this dialect" and if the dialect is registered and the interface for. Without it, we don't. Specifically, the way we do that right now is by having a DialectInterface where we say how to lower the dialect

I found another way to do this that is maybe more palatable? We can load the IREE dialect as part of construction of the other one, e.g.:

```c++
TFStringsDialect::TFStringsDialect(MLIRContext* context)
    : Dialect(getDialectNamespace(), context, TypeID::get<TFStringsDialect>()) {
  addInterfaces<TfStringsToHALConversionInterface>();
  addTypes<StringType>();
  context->getOrLoadDialect<IREE::Strings::StringsDialect>();

#define GET_OP_LIST
  addOperations<
#include "third_party/iree/integrations/tensorflow/compiler/dialect/tf_strings/ir/ops.cpp.inc"
      >();
}
```

We could also put this as a static method on the conversion interface:

```
TFStringsDialect::TFStringsDialect(MLIRContext* context)
    : Dialect(getDialectNamespace(), context, TypeID::get<TFStringsDialect>()) {
  addInterfaces<TfStringsToHALConversionInterface>();
  addTypes<StringType>();
  DialectRegistry registry;
  TfStringsToHALConversionInterface::getDependentDialects(registry);
  registry.loadAll(context);

#define GET_OP_LIST
  addOperations<
#include "third_party/iree/integrations/tensorflow/compiler/dialect/tf_strings/ir/ops.cpp.inc"
      >();
}
```

I'm also playing with trying to make it just happen on construction of the interface, but it's not working for some reason. I think this will work for us?

Reply:
This loads the dialect unnecessarily though even if the pass isn't called. What would be nice is for some way for this pass's `getDependentDialects` to look at the currently loaded dialects, but we can't here.

Reply:
> Right now with the global registry, we have a way for a out-of-tree user to say "here's how to lower this dialect" and if the dialect is registered and the interface for. Without it, we don't. 

Sure, but another way to see it that your extension mechanism for users to plug in your system was "hijacking" the upstream global registry in `ConvertFlowToHAL.cpp`:

```
    for (auto *dialect : context->getLoadedDialects()) {
```

There are many ways to get around this, for example a `IREEFlowToHalRegistry` wouldn't be flawed in the same way and would allow any out-of-tree user to plug themselves there.

Similarly, every extension points where you hit this issue is just a place where the IREE extension mechanism needs to be revisited and be more defensive (rely less on global side effects that you don't have control on, remove implicitness, etc.).

Reply:
Ok here's another way that I think is a lot better. Subclasses of HALConversionDialectInterface are responsible for loading dialects they need on construction. It's still not super pretty because there isn't an easy hook for them to implement (calling child methods from constructors is hard), and it sometimes loads the dialects unnecessarily, but at least it doesn't have layering issues.

Reply:
Btw, independently, I made post related to this here: https://llvm.discourse.group/t/rfc-revamp-dialect-registration/1559/3

I think it's plausible for a dialect to know all dialects that might be created by any of its interfaces or any interfaces implemented on its ops since the set of interfaces implemented by a dialect is a closed world fixed at dialect build time. Thus, I think that having `context->getOrLoadDialect<IREE::Strings::StringsDialect>();` in the dialect constructor is actually pretty palatable. Maybe the dependency on IREE::Strings::StringsDialect can be squirreled away in TfStringsToHALConversionInterface::getDependentDialects, but that seems minor. I don't see a strong need to go full lazy and only have IREE::Strings::StringsDialect loaded when TfStringsToHALConversionInterface is loaded, but it seems like a fine design as well -- we will need to be sure to collect all the interfaces in the FlowToHal conversion pass constructor, but that seems layered nicely.

One way this could be handled nicely by the framework is for DialectInterfaces to have a hook "getDependentDialects". addInterfaces<MyDialectInterface> in the dialect constructor would automatically call.

On the topic of
```
for (auto *dialect : context->getLoadedDialects()) {
```
I think it is actually layered nicely and is not a hack. Because we maintain the invariant "if there is an op in the module, then its dialect is loaded", then the above is just a handy alternative to this:
```
std::set<Dialect*> dialectSet;
module.walk([&](Operation *op) {
  dialectSet.insert(op->getDialect());
};
for (auto *dialect : dialectSet) {
```
Iterating over getLoadedDialects has the advantage of working even without access to the module. It has the disadvantage that it might iterate over more dialects than strictly necessary, but that seems generally ok.

Reply:
> On the topic of
> for (auto *dialect : context->getLoadedDialects()) {
> I think it is actually layered nicely and is not a hack. Because we maintain the invariant "if there is an op in the module, then its dialect is loaded", then the above is just a handy alternative to this:

It is nice in principle but:
1) That works only if you already have the module, you can't create the pipeline first and then get a module. You'd have to recreate the pipeline for every new module.
2) A pass in the pipeline can introduce a new dialect, which will be loaded through `getDependentDialects()` at the "run()" time, while you seem to need the information at pipeline build time here.